modules/setting/storage.go

```diff
@@ -91,134 +91,161 @@ func getStorage(rootCfg ConfigProvider, name, typ string, sec ConfigSection) (*S
 		return nil, errors.New("no name for storage")
 	}
 
-	var targetSec ConfigSection
-	// check typ first
-	if typ != "" {
-		var err error
-		targetSec, err = rootCfg.GetSection(storageSectionName + "." + typ)
-		if err != nil {
-			if !IsValidStorageType(StorageType(typ)) {
-				return nil, fmt.Errorf("get section via storage type %q failed: %v", typ, err)
-			}
+	targetSec, tp, err := getStorageTargetSection(rootCfg, name, typ, sec)
+	if err != nil {
+		return nil, err
+	}
+
+	overrideSec := getStorageOverrideSection(rootCfg, targetSec, sec, tp, name)
+
+	targetType := targetSec.Key("STORAGE_TYPE").String()
+	switch targetType {
+	case string(LocalStorageType):
+		return getStorageForLocal(targetSec, overrideSec, tp, name)
+	case string(MinioStorageType):
+		return getStorageForMinio(targetSec, overrideSec, tp, name)
+	default:
+		return nil, fmt.Errorf("unsupported storage type %q", targetType)
+	}
+}
+
+const (
+	targetSecIsTyp             = iota // target section is [storage.type] which the type from parameter
+	targetSecIsStorage                // target section is [storage]
+	targetSecIsDefault                // target section is the default value
+	targetSecIsStorageWithName        // target section is [storage.name]
+	targetSecIsSec                    // target section is from the name seciont [name]
+)
+
+func getStorageSectionByType(rootCfg ConfigProvider, typ string) (ConfigSection, int, error) {
+	targetSec, err := rootCfg.GetSection(storageSectionName + "." + typ)
+	if err != nil {
+		if !IsValidStorageType(StorageType(typ)) {
+			return nil, 0, fmt.Errorf("get section via storage type %q failed: %v", typ, err)
 		}
-		if targetSec != nil {
-			targetType := targetSec.Key("STORAGE_TYPE").String()
-			if targetType == "" {
-				if !IsValidStorageType(StorageType(typ)) {
-					return nil, fmt.Errorf("unknow storage type %q", typ)
-				}
-				targetSec.Key("STORAGE_TYPE").SetValue(typ)
-			} else if !IsValidStorageType(StorageType(targetType)) {
-				return nil, fmt.Errorf("unknow storage type %q for section storage.%v", targetType, typ)
-			}
+	}
+	if targetSec == nil {
+		return nil, 0, nil
+	}
```

Comment:
I am pretty confused by this logic.
With your current approach, you allow querying `storage.local`/`storage.minio`, something that has never been documented to be possible.
Is it supposed to be possible? Or is it simply an oversight at the moment?
I think the logic is supposed to be something like
`validStorageType(typ) => return nil, 0, nil;`
`target := section(name + "." + typ)`
`validStorageType(target) => return targetSec, targetSecIsTyp, nil`
otherwise
`error`

Reply:
I did some changes for this one. Please review it again.